### PR TITLE
Rechunk array if smallest chunk size is smaller than depth

### LINKED ISSUE #6597: map_overlap fails if last chunk is smaller than depth
**What happened**:

`dask.array.map_overlap` throws a `ValueError` if the last chunk is smaller than the `depth` parameter. (The size of the last chunk is normally outside the user's control.)

**What you expected to happen**:

The function should work in this case. The last chunk should really be ignored for this check since it can be any size less than the array chunksize.

**Minimal Complete Verifiable Example**:

In this example the last chunk is of size 1, and the depth is 2:

```python
>>> import dask.array as da
>>> import numpy as np
>>> x = da.from_array(np.arange(11), chunks=5)
>>> y = x.map_overlap(lambda x: x + len(x), depth=2, boundary=0).compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tom/workspace/dask/dask/array/core.py", line 2240, in map_overlap
    func, self, depth=depth, boundary=boundary, trim=trim, **kwargs
  File "/Users/tom/workspace/dask/dask/array/overlap.py", line 716, in map_overlap
    args = [overlap(x, depth=d, boundary=b) for x, d, b in zip(args, depth, boundary)]
  File "/Users/tom/workspace/dask/dask/array/overlap.py", line 716, in <listcomp>
    args = [overlap(x, depth=d, boundary=b) for x, d, b in zip(args, depth, boundary)]
  File "/Users/tom/workspace/dask/dask/array/overlap.py", line 489, in overlap
    "more evenly divides the shape of your array." % (d, min(c))
ValueError: The overlapping depth 2 is larger than your
smallest chunk size 1. Rechunk your array
with a larger chunk size or a chunk size that
more evenly divides the shape of your array.
```

`y` should be `9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 15]`

**Anything else we need to know?**:

**Environment**:

- Dask version: 2.25.0
- Python version: 3.7.6
- Operating System: Mac 10.13.6
- Install method (conda, pip, source): source

### LINKED ISSUE #6688: map_overlap should work with smaller-than-depth chunk sizes
Related to #6597, but distinct I think.

If I have an array chunked into individual slices along an axis, but I want to perform an operation across that axis, I would *expect* `map_overlap` to be able to pull as many chunks as it needs along that axis to make the boundary, ie this should work:

```python
from dask import array as da
from scipy import ndimage as ndi

rand = da.random.random((860, 1024, 1024), chunks=(1, 1024, 1024))
filtered = da.map_overlap(
    rand,
    lambda arr: ndi.median_filter(arr, 5),
    depth=(2, 2, 2),
    boundary='reflect',
)
```

This raises:

```pytb
ValueError: The overlapping depth 2 is larger than your
smallest chunk size 1. Rechunk your array
with a larger chunk size or a chunk size that
more evenly divides the shape of your array.
```

But one *should* be able to grab multiple chunks to pad sufficiently?

If auto-rechunking is the preferred solution, then feel free to close this, since I think #6597 covers that adequately.

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`
- [x] closes #6597,  closes #6688

### CODE REVIEW

**Comment by jsignell** [comment]
This is ready for review @dask/maintenance

File: dask/array/overlap.py
Comment (mrocklin) [inline dask/array/overlap.py:510]: I wonder if this might be simpler if we handle it on a per-axis basis.

```python
>>> ensure_minimum_chunksize(size=10, chunks=(20, 5, 100, 25, 5))
(25, 100, 30)
```

And then we could call this on each of the chunks axes

```python
new_chunks = tuple(ensure_minimum_chunksize(depth, cs) for depth.values(), chunks)
if new_chunks != chunks:
    x = x.rechunk(new_chunks)
```

This might require that we normalize things like depth ahead of time, but that normalization should probaby be separate from this rechunking process.

File: dask/array/overlap.py
Comment (mrocklin) [inline dask/array/overlap.py:496]: I'm not sure that this function is what we want here.  We don't care about chunks being a multiple of a number, we care about them being at least a certain size

**Comment by chrisroat** [comment]
I really like the idea of the dask framework handling the chunking issues under the hood.

One suggestion I may offer, and I could do in a follow-up if desired, is to allow an option for the small chunk to steal from a neighboring chunk. 

I ran into this small chunk issue in the past, and tried to combine chunks like this PR.  It worked fine for most cases, but then I ran into an odd edge case.  My operation was using a library that was padding out my chunks dimensions to a power of 2 before loading it onto a GPU.  Using combined chunks, a corner chunk of a 3-d array was formed that was went over the power-of-2 threshold in all dimensions, leading to an 8x increase in size of the padded array.  Needless to day, I hadn't planned for that possibility.

File: dask/array/overlap.py
Comment (jsignell) [reply dask/array/overlap.py]: Thank you for looking at this PR! I think it's a lot more straightforward now.

**Comment by jsignell** [comment]
> One suggestion I may offer, and I could do in a follow-up if desired, is to allow an option for the small chunk to steal from a neighboring chunk.

I'm not quite sure I understand. Are you suggesting that rather than combining small chunks with neighboring big chunks, we should steal bits of big chunks to add on to small chunks? That seems like a fair amount of extra complexity.

**Comment by chrisroat** [comment]
> > One suggestion I may offer, and I could do in a follow-up if desired, is to allow an option for the small chunk to steal from a neighboring chunk.
> 
> I'm not quite sure I understand. Are you suggesting that rather than combining small chunks with neighboring big chunks, we should steal bits of big chunks to add on to small chunks? That seems like a fair amount of extra complexity.

The extra work could be worth it.  If someone has chunked data to a certain size, there could be a could reason.  A rechunking that makes bigger chunks will surprise a lot of people.  It can cause blow past the memory limits of a worker  --  this is exactly what happened to me.

**Comment by jsignell** [comment]
Thanks for the idea @chrisroat! I implemented it and I think it's pretty neat!

File: dask/array/tests/test_overlap.py
Comment (chrisroat) [inline dask/array/tests/test_overlap.py:767]: I would suggest adding a case similar to `test_depth_greater_than_smallest_chunk_combines_chunks`, where concatenation does have to happen, and it creates chunks larger than any in the original, oi.e.
(10, 10, 1) -> (10, 11)

File: dask/array/tests/test_overlap.py
Comment (chrisroat) [inline dask/array/tests/test_overlap.py:737]: Duplicate name with previous test.

File: dask/array/tests/test_overlap.py
Comment (chrisroat) [inline dask/array/tests/test_overlap.py:619]: Combining chunks seems clean, but I could see an argument for these to be rechunked into (prior to padding): 
(4,4,4) and (10, 2)

**Comment by chrisroat** [comment]
Whoa..  and not a lot of additional complexity.  The `test_ensure_minimum_chunksize` tests are very helpful to see your algorithm in action.

File: dask/array/tests/test_overlap.py
Comment (jsignell) [reply dask/array/tests/test_overlap.py:737]: thanks! I think I didn't mean to have  one of those

File: dask/array/tests/test_overlap.py
Comment (jsignell) [reply dask/array/tests/test_overlap.py:619]: I think I'd rather not get into that kind of rechunking since that is very easy for people to do their own selves.

File: dask/array/tests/test_overlap.py
Comment (jsignell) [reply dask/array/tests/test_overlap.py:767]: good idea!

File: dask/array/tests/test_overlap.py
Comment (mrocklin) [inline dask/array/tests/test_overlap.py:769]: Nice test!

File: dask/array/tests/test_overlap.py
Comment (mrocklin) [inline dask/array/tests/test_overlap.py:759]: Black is behaving oddly here?

File: dask/array/tests/test_overlap.py
Comment (mrocklin) [inline dask/array/tests/test_overlap.py:734]: assert_eq?

File: dask/array/tests/test_overlap.py
Comment (mrocklin) [inline dask/array/tests/test_overlap.py:733]: I think that this more clearly  defines what we care about:

```suggestion
    assert all(c >= 2 for c in y.chunks)
```

That way if someone changes the policy in the future this test doesn't gum up their activity.  We really only care that chunks are at least the size of the depth.  We don't want to be too perscriptive about the exact chunking.  My experience is that being too perscriptive results in developers spending time "fixing up the tests" in future work

File: dask/array/tests/test_overlap.py
Comment (mrocklin) [inline dask/array/tests/test_overlap.py:625]: I'm not sure that we want to assert this here

Comment (mrocklin) [reply dask/array/tests/test_overlap.py:625]: Same elsewhere

File: dask/array/overlap.py
Comment (mrocklin) [inline dask/array/overlap.py:520]: I think that `rechunk` may already have the if/else behavior.  If not then let's add it there rather than here.

```python
assert x.rechunk(x.chunks) is x
```

File: dask/array/overlap.py
Comment (mrocklin) [inline dask/array/overlap.py:426]: With finicky algorithmic functions like this I find that having one simple example in the docstring does wonders for future maintainers.  Even though it's internal and users won't see it, seeing a quick docstring example can be a great accelerant to understanding for the next person who has to deal with the overlap code.

**Review by mrocklin — APPROVED** [review]
Thanks @jsignell !

In principle the implementation here seems sound to me.  I made a few finicky remarks around testing.  They're not critical, but they seem like maybe good habits to adopt.

File: dask/array/tests/test_overlap.py
Comment (jsignell) [reply dask/array/tests/test_overlap.py:753]: Yeah I probably had a trailing comma and it got over zealous.

**Comment by jsignell** [comment]
Ok thanks for the second review @mrocklin! I'll merge this when green.

File: dask/array/tests/test_overlap.py
Comment (martindurant) [reply dask/array/tests/test_overlap.py:753]: I believe different versions of black have different opinions on trailing (pseudo-Oxford?) commas.

**Pull request merged**
